### PR TITLE
random operations test

### PR DESCRIPTION
**Pull request by a-masterov**
## Problem
We need to test the stability of Neon.

## Summary of changes
The test runs random operations on a Neon project. It performs via the Public API calls the following operations: `create a branch`, `delete a branch`, `add a read-only endpoint`, `delete a read-only endpoint`, `restore a branch to a random position in the past`. All the branches and endpoints are loaded with `pgbench`.

### CODE REVIEW

File: test_runner/fixtures/neon_api.py [hunk 1]
```diff
@@ -34,6 +34,14 @@ def __request(self, method: str | bytes, endpoint: str, **kwargs: Any) -> reques
...
         resp = requests.request(method, f"{self.__neon_api_base_url}{endpoint}", **kwargs)
         log.debug("%s %s returned a %d: %s", method, endpoint, resp.status_code, resp.text)
+        if resp.status_code >= 400:
+            log.error(
+                "%s %s returned a %d: %s",
+                method,
+                endpoint,
+                resp.status_code,
+                resp.text if resp.status_code != 524 else "CloudFlare error page",
+            )
```
Comment (Bodobolero) [inline test_runner/fixtures/neon_api.py:44]:
```suggestion
        if resp.status_code >= 400:
            log.error(
                "%s %s returned a %d: %s",
                method,
                endpoint,
                resp.status_code,
                resp.text if resp.status_code != 524 else "CloudFlare error page",
            )
        else:
            log.debug("%s %s returned a %d: %s", method, endpoint, resp.status_code, resp.text)
```

File: test_runner/fixtures/neon_api.py [hunk 2]
```diff
@@ -101,6 +109,90 @@ def delete_project(
 
         return cast("dict[str, Any]", resp.json())
 
+    def create_branch(
+        self,
+        project_id: str,
+        branch_name: str | None = None,
+        parent_id: str | None = None,
+        add_endpoint=True,
```
Comment (Bodobolero) [inline test_runner/fixtures/neon_api.py:117]: to make this function more generic and usable in the future I suggest to support more parameters from public API: https://api-docs.neon.tech/reference/createprojectbranch

Specifically:
- parent_lsn
- parent_timestamp
- protected
- archived
- init_source

File: test_runner/fixtures/neon_api.py (hunk 1 above)
Comment (a-masterov) [reply test_runner/fixtures/neon_api.py:44]: Good catch, thanks

File: test_runner/fixtures/neon_api.py [hunk 3]
```diff
@@ -101,6 +110,90 @@ def delete_project(
 
         return cast("dict[str, Any]", resp.json())
```
Comment (Bodobolero) [inline test_runner/fixtures/neon_api.py:112]: You added some API requests to this fixture, but
https://api-docs.neon.tech/reference/

has a lot of additional APIs.
What was the decision criteria for subset selection?
Did you use grafana dashboard to verify which API endpoints are most frequently called (that data is available)?

File: test_runner/random_ops/test_random_ops.py [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+        target_object = self.endpoints[target] if is_endpoint else self.branches[target]
+        if target_object.connect_env is None:
+            raise RuntimeError(f"The connection environment is not defined for {target}")
+        log.info(
+            "running pgbench on %s, cmd: %s, host: %s",
+            target,
+            cmd,
+            target_object.connect_env["PGHOST"],
+        )
+        pgbench = self.pg_bin.run_nonblocking(
```
Comment (Bodobolero) [inline test_runner/random_ops/test_random_ops.py:315]: Here we should make sure the compute can be connected to within a certain time threshold (e.g. at max 5 seconds) and report everything else as a failure, too

**Review by Bodobolero — APPROVED** [review]
I added a few questions and suggestions.
Coverage of additional APIs (as suggested below) can be moved to separate PR if you want, or you can add them in this PR.

File: test_runner/random_ops/test_random_ops.py (hunk 4 above)
Comment (a-masterov) [reply test_runner/random_ops/test_random_ops.py:315]: We check if all the benchmarks are alive after every operation, so if a benchmark unexpectedly dies, it will be detected and reported as a failure.

File: test_runner/fixtures/neon_api.py (hunk 3 above)
Comment (a-masterov) [reply test_runner/fixtures/neon_api.py:112]: It's interesting, but we have to create the data dict manually anyway.

File: test_runner/random_ops/test_random_ops.py (hunk 4 above)
Comment (Bodobolero) [reply test_runner/random_ops/test_random_ops.py:315]: and report an error if it times out or fails

File: test_runner/fixtures/neon_api.py (hunk 3 above)
Comment (a-masterov) [reply test_runner/fixtures/neon_api.py:112]: I believe, we can easily add the required function to `neon_api.py` when we need it.

File: test_runner/random_ops/test_random_ops.py [hunk 5]
```diff
@@ -0,0 +1,472 @@
...
+        else:
+            log.info("Leaf branches not found, skipping")
+    elif action == "new_ro_endpoint":
+        ep = random.choice(
+            [br for br in project.branches.values() if br.id not in project.reset_branches]
+        ).create_ro_endpoint()
+        log.info("Created the RO endpoint with id %s branch: %s", ep.id, ep.branch.id)
+        ep.start_benchmark()
+    elif action == "delete_ro_endpoint":
+        ro_endpoints = [_ for _ in project.endpoints.values() if _.type == "read_only"]
```
Comment (bayandin) [inline test_runner/random_ops/test_random_ops.py:414]:
```suggestion
        ro_endpoints = [ep for ep in project.endpoints.values() if ep.type == "read_only"]
```

File: test_runner/random_ops/test_random_ops.py [hunk 6]
```diff
@@ -0,0 +1,472 @@
...
+        rc = self.benchmarks[target].poll()
+        if rc is not None:
+            _, err = self.benchmarks[target].communicate()
+            log.error("STDERR: %s", err)
+            # if the benchmark failed due to irresponsible Control plane,
+            # just restart it
+            if (
+                "ERROR:  Couldn't connect to compute node" in err
+                or "ERROR:  Console request failed" in err
+            ):
```
Comment (bayandin) [inline test_runner/random_ops/test_random_ops.py:343]: These errors look too broad and might cover pretty different underlying issues.
Let's fail the test at the end if any of such errors appear during the test (otherwise we might miss all the errors)

File: .github/workflows/random-ops-test.yml [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+        description: 'The random seed'
+        required: false
+        default: 0
+      num_operations:
+        type: number
+        description: "The number of operations to test"
+        default: 250
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref_name }}
```
Comment (bayandin) [inline .github/workflows/random-ops-test.yml:25]: Probably we don't need it, test doesn't use any resources that should be locked, and each new test run different from the previous one

File: .github/workflows/random-ops-test.yml [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+on:
+  schedule:
+    # * is a special character in YAML so you have to quote this string
+    #          ┌───────────── minute (0 - 59)
+    #          │ ┌───────────── hour (0 - 23)
+    #          │ │ ┌───────────── day of the month (1 - 31)
+    #          │ │ │ ┌───────────── month (1 - 12 or JAN-DEC)
+    #          │ │ │ │ ┌───────────── day of the week (0 - 6 or SUN-SAT)
+    - cron:  '23 02 * * *' # run once a day, timezone is utc
```
Comment (bayandin) [inline .github/workflows/random-ops-test.yml:11]: I think we want to run it more frequently, if it takes 20m, let's say every hour or two?

File: .github/workflows/random-ops-test.yml [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref_name }}
+
+defaults:
+  run:
+    shell: bash -euxo pipefail {0}
+
+permissions:
+  id-token: write # aws-actions/configure-aws-credentials
+  statuses: write # require for posting a status update
```
Comment (bayandin) [inline .github/workflows/random-ops-test.yml:33]: Let's give permission only for the job, and use empty permission by default, like here: https://github.com/neondatabase/neon/blob/53f54ba37aa58f99cc9fa2966fbecdc06419cfb7/.github/workflows/build-build-tools-image.yml#L39-L40

File: .github/workflows/random-ops-test.yml [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+  DEFAULT_PG_VERSION: 16
+  PLATFORM: neon-captest-new
+  AWS_DEFAULT_REGION: eu-central-1
+
+jobs:
+  check-permissions:
+    if: ${{ !contains(github.event.pull_request.labels.*.name, 'run-no-ci') }}
+    uses: ./.github/workflows/check-permissions.yml
+    with:
+      github-event-name: ${{ github.event_name }}
```
Comment (bayandin) [inline .github/workflows/random-ops-test.yml:45]: This workflow can be triggered only manually, I don't think we need to check permissions for that (or label)

File: .github/workflows/random-ops-test.yml [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+    strategy:
+      fail-fast: false
+      matrix:
+        pg-version: [16, 17]
+
+    container:
+      image: ${{ needs.build-build-tools-image.outputs.image }}-bookworm
+      credentials:
+        username: ${{ secrets.NEON_DOCKERHUB_USERNAME }}
+        password: ${{ secrets.NEON_DOCKERHUB_PASSWORD }}
```
Comment (bayandin) [inline .github/workflows/random-ops-test.yml:65]: We recently switched to GHCR, need to update credentials: https://github.com/neondatabase/neon/blob/53f54ba37aa58f99cc9fa2966fbecdc06419cfb7/.github/workflows/pg-clients.yml#L54-L57

File: .github/workflows/random-ops-test.yml [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+      options: --init --user root
```
Comment (bayandin) [inline .github/workflows/random-ops-test.yml:66]: I don't think we need `--user root` here, we don't use docker

**Review by bayandin — COMMENTED** [review]
Great progress, so far!

File: test_runner/random_ops/README.md [hunk 13]
```diff
@@ -0,0 +1,93 @@
+# Random Operations Test for Neon API Stability
+
+## Problem Statement
+
+Neon needs robust testing of its API stability to ensure reliability for users. The random operations test addresses this by continuously exercising the API with unpredictable sequences of operations, helping to identify edge cases and potential issues that might not be caught by deterministic tests.
```
[the remark this answers is not in the document]
Comment (a-masterov) [reply test_runner/random_ops/README.md]: Rephrased

File: .github/workflows/random-ops-test.yml (hunk 10 above)
Comment (a-masterov) [reply .github/workflows/random-ops-test.yml:45]: Good catch, removed

File: .github/workflows/random-ops-test.yml (hunk 12 above)
Comment (a-masterov) [reply .github/workflows/random-ops-test.yml]: Removed, thanks

Comment (a-masterov) [reply .github/workflows/random-ops-test.yml]: Removing `--user root` causes the following error
```
/usr/bin/docker exec  151d9ec0af3cfd9dada52557b68e514d49409975d0c6382bc35704d64f3da188 sh -c "cat /etc/*release | grep ^ID"
node:fs:2367
    return binding.writeFileUtf8(
                   ^
Error: EACCES: permission denied, open '/__w/_temp/_runner_file_commands/save_state_5d60a47f-f24d-4d09-babd-8352525d83b0'
```
[Run](https://github.com/neondatabase/neon/actions/runs/13988162196/job/39165951729)

File: .github/workflows/random-ops-test.yml (hunk 11 above)
Comment (a-masterov) [reply .github/workflows/random-ops-test.yml]: It produces the following error:
```
/usr/bin/docker --config /home/runner/work/_temp/.docker_89df6f9f-7a60-45d6-bd3e-46e82d2c1884 login  -u a-masterov --password-stdin
Warning: Docker login for '' failed with exit code 1, back off 7.105 seconds before retry.
/usr/bin/docker --config /home/runner/work/_temp/.docker_89df6f9f-7a60-45d6-bd3e-46e82d2c1884 login  -u a-masterov --password-stdin
Error: Docker login for '' failed with exit code 1
```
[Run](https://github.com/neondatabase/neon/actions/runs/13988481576/job/39166909375)

File: test_runner/random_ops/test_random_ops.py (hunk 6 above)
Comment (a-masterov) [reply test_runner/random_ops/test_random_ops.py:343]: We can only use the error messages from stderr to find out the cause of the failure. `Console request failed` can definitely tell us that the Console was the cause.

I believe if we want to fail after such failures, it's better to fail immediately after an error to allow easily find the problem. Or we can use `::warning` to highlight the issues, but not fail.

File: .github/workflows/random-ops-test.yml (hunk 11 above)
Comment (a-masterov) [reply .github/workflows/random-ops-test.yml]: Fixed image name, now it works

Comment (a-masterov) [reply .github/workflows/random-ops-test.yml]: It's done now

File: .github/workflows/random-ops-test.yml [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+      fail-fast: false
+      matrix:
+        pg-version: [16, 17]
+
+    container:
+      image: ghcr.io/neondatabase/build-tools:pinned-bookworm
+      credentials:
+        username: ${{ github.actor }}
+        password: ${{ secrets.GITHUB_TOKEN }}
+      options: --init --user root
```
Comment (bayandin) [inline .github/workflows/random-ops-test.yml:52]: Let's try `nonroot`:

```suggestion
      options: --init --user nonroot
```

File: .github/workflows/random-ops-test.yml (hunk 12 above)
Comment (a-masterov) [reply .github/workflows/random-ops-test.yml]: I've changed `runs-on` to `small`, and it runs now without `--root`

File: .github/workflows/random-ops-test.yml (hunk 14 above)
Comment (a-masterov) [reply .github/workflows/random-ops-test.yml]: Changing the runner fixed this.

**Review by bayandin — APPROVED** [review]
🚀 🚀 🚀
